tools/scan_includes.c

```diff
@@ -42,29 +42,49 @@ void scan_file(const char *filename, bool strict) {
 
 	for (char *ptr = contents; ptr && ptr - contents < size; ptr++) {
 		bool is_incbin = false, is_include = false;
+		ptr = strpbrk(ptr, ";\"Ii");
+		if (!ptr) {
+			break;
+		}
 		switch (*ptr) {
 		case ';':
-			ptr = strchr(ptr, '\n');
+			// Skip comments until the end of the line
+			ptr = strchr(ptr + 1, '\n');
 			if (!ptr) {
 				fprintf(stderr, "%s: no newline at end of file\n", filename);
+				goto done;
 			}
 			break;
 		case '"':
-			ptr++;
-			ptr = strchr(ptr, '"');
+			ptr = strchr(ptr + 1, '"');
 			if (ptr) {
 				ptr++;
 			} else {
 				fprintf(stderr, "%s: unterminated string\n", filename);
+				goto done;
 			}
 			break;
 		case 'I':
 		case 'i':
+			// Check that an INCLUDE/INCBIN starts as its own token
+			char before = ptr > contents ? *(ptr - 1) : '\n';
+			if (before != ' ' && before != '\t' && before != '\n' && before != ':') {
+				break;
+			}
 			is_incbin = !strncmp(ptr, "INCBIN", 6) || !strncmp(ptr, "incbin", 6);
 			is_include = !strncmp(ptr, "INCLUDE", 7) || !strncmp(ptr, "include", 7);
 			if (is_incbin || is_include) {
-				ptr = strchr(ptr, '"');
-				if (ptr) {
+				// Check that an INCLUDE/INCBIN ends as its own token
+				char after = is_include ? *(ptr + 7) : *(ptr + 6);
+				if (after != ' ' && after != '\t' && after != '"') {
+					break;
+				}
+				ptr = strpbrk(ptr, "\"\n");
```

Comment:
After this, it will be the case that `*ptr == '"'` if you have an `INCLUDE` without a file path followed by a newline, i.e. it's not the end of the file. Then shouldn't it print the "no file path" warning message, and continue processing more lines? Basically rewrite the code below to this? :

```c
if (*ptr == '"') {
  // the code to do the recursive scan_file
} else {
  fprintf(stderr, "%s: no file path after INC%s\n", filename, is_include ? "LUDE" : "BIN");
  if (!ptr) {
    goto done;
  }
}
```